### PR TITLE
Fix bug in sighash for coinbase transactions

### LINKED ISSUE #1939: Fix sapling binding signature errors
## Motivation

In #1377, we discovered that Zebra's sapling binding signature verification fails on valid transactions.

## Solution

Diagnose the errors in the binding signature code, and re-enable async binding signature validation:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-consensus/src/transaction.rs#L271

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

The sighash of coinbase transactions was being computed incorrectly because it ignored the "outpoint" of coinbase inputs while they shouldn't be.

Closes #1939.

### Specifications

https://zips.z.cash/zip-0143, which does not have any exceptions for the coinbase transaction. (It could be more explicit about it...)
https://developer.bitcoin.org/reference/transactions.html#coinbase-input-the-input-of-the-first-transaction-in-a-block describes the "outpoint" of a coinbase transaction (a null hash, and a 0xFFFFFFFF index).

### Designs

N/A

## Solution

Recreate the "fake" outpoint of coinbase inputs in order to hash it.

I added two tests:

- One test was based on @dconnolly's branch cited in #1939 which iterates the test blocks and verifies the Sapling binding signature for each transaction where applicable. This reproduces the bug.
- One test was added based on ZIP-243 test vectors from https://github.com/zcash-hackworks/zcash-test-vectors/ . This does _not_ reproduce the bug since it generates random hashes, which are almost impossible to be all-zero (which would emulate a coinbase transaction and trigger the bug). I added it while investigating and I thought it wouldn't hurt to be added. (It will also be useful for #2183)

Note: our intention, as discussed in Discord, was to replace the entire sighash with librustzcash implementation. However I ended up identifying the bug while preparing the tests the replacement. We'll probably end up going forward with this strategy (#2183), but I wanted to fix the bug before doing that (and deleting the zebra sighash code) so that if for whatever reason we decide one day to go back to our own implementation, it won't have this bug.

## Review

See #1939 for what this is blocking.

I think anyone can review this.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

We'll probably follow this up with #2183

### CODE REVIEW

**Comment by dconnolly** [comment]
> https://zips.z.cash/zip-0143, which does not have any exceptions for the coinbase transaction. (It could be more explicit about it...)

All the signature hash zips miss specifying this coinbase detail, and all the test vectors in zips and in zcash-hackworks fail to exercise it, and the new ZIP-244 non-malleable tx id mentions coinbase but not relevant to this bug. I'm not sure exactly where this should be clarified but it is underspecified and under-tested to achieve consensus from the specs.

**Comment by teor2345** [comment]
@conradoplg can you open a ticket in https://github.com/zcash/zips to specify the coinbase output sighash?

There's no need to open a PR, just describing the spec gap is enough.

Edit:
Let's also list ZIP-244 in the bug, because it also doesn't specify what happens to coinbase inputs:
https://zips.z.cash/zip-0244#t-2a-prevouts-digest

File: zebra-chain/src/transaction/sighash.rs [hunk 1]
```diff
@@ -183,9 +183,12 @@ impl<'a> SigHasher<'a> {
...
             .iter()
-            .filter_map(|input| match input {
-                transparent::Input::PrevOut { outpoint, .. } => Some(outpoint),
-                transparent::Input::Coinbase { .. } => None,
+            .map(|input| match input {
+                transparent::Input::PrevOut { outpoint, .. } => outpoint,
+                transparent::Input::Coinbase { .. } => &transparent::OutPoint {
+                    hash: crate::transaction::Hash([0; 32]),
+                    index: u32::MAX,
+                },
```
Comment (teor2345) [inline zebra-chain/src/transaction/sighash.rs:191]: Yeah that's not specified in any of the relevant ZIPs.

**Review by teor2345 — DISMISSED** [review]
Let's merge this after we've opened the ZIP bug ticket.

Then we can open another PR to switch to `librustzcash` sighash and delete this code.

(Please feel free to dismiss my review once the ticket is open.)

**Comment by conradoplg** [comment]
> @conradoplg can you open a ticket in https://github.com/zcash/zips to specify the coinbase output sighash?

Opened in https://github.com/zcash/zips/issues/535

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**